enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+* It uses [efs-utils](https://github.com/aws/efs-utils) to provide encryption in transport, which is not included in
+  RHEL. We will need to package it, either as an image or RPM package.
+
+## Proposal
+High level only, see design details below.
+
+1. Rewrite current aws-efs-operator to use [CSI driver installation functions of library-go](https://github.com/openshift/library-go/tree/master/pkg/operator/csi/csicontrollerset).
+   * This allows us to share well tested code to install CSI drivers and implement bugfixes / new features at a single
+     place.
+   * It will add common features of CSI driver operators which are missing in the current community operator, such as
+     proxy injection and restarting the drivers when cloud credentials change.
+
+2. Update (rewrite) current aws-efs-operator to watch `ClusterCSIDriver` CR named `efs.csi.aws.com`.
+   * `ClusterCSIDriver` CRD is already provided by OCP and is used to install all other CSI drivers that OCP ships.
+   * Users must explicitly create the CR to install the CSI driver to allow users to uninstall the driver if they don't
+     like it.
+
+3. Keep existing functionality:
+   * `SharedVolume` CRD in `aws-efs.managed.openshift.io/v1alpha1` API group keeps working as it is now.
+   * Update from the community operator to the supported operator should be as seamless as possible.
+   * The operator is still installed via OLM.
+
+4. Ship the operator + CSI driver through ART pipeline.
+   * Most of the operator just installs / removes CSI driver, i.e. manages driver's RBAC rules, CredentialsRequest,
+     DaemonSet, Deployment, CSIDriver and StorageClasses. See existing
+     [AWS EBS operator](https://github.com/openshift/aws-ebs-csi-driver-operator/tree/master/assets) for objects it
+     manages, EFS should be very similar in this regard.
+   * `SharedVolume` processing on top of it is very trivial, it just creates PV + PVC.
+
+5. Ship [efs-utils](https://github.com/aws/efs-utils) as a base image for the EFS CSI driver image.
+
+6. Update installer to delete Access Points and EFS volumes that have `owner` tag when destroying a cluster.
+   * Existence of EFS volume using a VPC / Subnets actively prevents installer to delete them.
+   * We should allow users to tag the volumes to be deleted when the cluster is deleted. At least CI should use this
+     approach to ensure there are no leftovers.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Currently, all cloud-based CSI drivers shipped by OCP are installed when a cluster is installed on the particular cloud.
+AWS EFS will not be installed by default. We will include a note about opt-in installation in docs.
+
+After installation of the new operators, either as a fresh install or upgrading the existing one, users must create the
+CR when the operator is installed on a new cluster. This is something that was not necessary in the previous releases of
```

Comment:
Need to see proposed changes to the CRD, if any, and an example of what a CR will look like to (de)activate the EFS CSI driver. (Not sure if that level of detail is appropriate for this document, but you included dockerfile snippets so.... :) )

Reply:
Here's the CR the cluster-storage-operator creates for EBS: https://github.com/openshift/cluster-storage-operator/blob/master/assets/csidriveroperators/aws-ebs/10_cr.yaml

Reply:
Added link to EBS CR.